Azure/sonic-buildimage
# [Services] Restart DHCP-Relay service upon unexpected critical process exit.

- What I did
Restart DHCP-Relay service if one of critical processes running in DHCP-Relay container exited or crashed abnormally.

- How I did it
Generally I follow the framework created by Joe to implement this feature in DHCP-Relay container.
**First**, add supervisor-proc-exit-listener event listener option in Supervisord configuration file in DHCP-Relay docker container. Supervisord will read a list of critical processes for which to monitor the unexpected crashed and exited. 
For DHCP-Relay container, since a bunch of critical processes will be monitored by a group, we only need put the groupname in the file critical_processes. At the same time, we also add source code in supervisor-proc-exit-listener script to retrieve the groupname and then decide whether it appears in  critical_processes.
**Second**, configure dhcp-relay.service to always auto-restart the service if it stops, with a delay of 30 seconds. Also set a rate limit of 3 restarts within 20 minutes (1200 seconds).

- How to verify it
On your switch device, please use _docker ps_ command to list all running docker containers.
Then use _docker exec -it container_id /bin/bash_ to login target container. Typing _top_ command
on the shell will display all the processes dynamically and you will spot the process id of one
of the critical processes. Finally type the command _kill -9 process_id_ to terminate one process.
After exiting the container, you can use _watch -n 1 docker ps_ to dynamically see the restart
of DHCP-Relay container.
